### ISSUE TITLE
Executing a method on the model after model binding

### ISSUE THREAD

**Issue by mrahhal** [aspnet/Mvc#5614]
Suppose that we have a simple action:

```cs
public IActionResult Foo([FromBody] FooModel model)
{
    ...
}
```

We can use `IValidatableObject` (by implementing it on the model) to run custom validation when the model is being validated. Is there a similar interface to implement in order to run a custom method on the model when model binding finishes?

The use case is to make simple changes to the model after model binding (for example trimming some strings among other things). Rather than having to call a method on the model in every action, having a custom method execute with model binding would be helpful and cleaner.

I looked at the code extensively, but couldn't find anything.

---

Now that I think about it, it might be possible to replace the default `BodyModelBinder` for a custom one and run the custom method at the end of model binding if the model implements a certain custom interface.

**Comment by rynowak** [comment]
This is a good use case for implementing an `IActionFilter` if you want this to run on model after modelbinding/validation but before the action.

If you want to run custom code after model binding but before validation then you'll need to implement a custom model binder.

**Comment by mrahhal** [comment]
@rynowak yeah it would work as an action filter. But it doesn't differ a lot from calling `model.Foo()` in each action (I know that I can have an action on the controller itself).

What I picture is having a method on the model itself that will be executed all the time whenever model binding does its work. It's a lot cleaner this way, and really it's the model's job to do those minor edits on its properties (like calling `ToLowerCase` on a certain property for example).

**Comment by rynowak** [comment]
> But it doesn't differ a lot from calling model.Foo() in each action (I know that I can have an action on the controller itself).

The difference is that you can apply an action filter globally. I'd do it something like this:
```
public interface INeedsCleanup
{
    void DoCleanup();
}

public class SomeModel : INeedsCleanup
{
   ....
}

public class MyCleanupFilter : IActionFilter
{
    public void OnActionExecuting(ActionExecutingContext context)
    {
        foreach (var kvp in context.ActionArguments)
        {
            (kvp.Value as INeedsCleanup)?.DoCleanup();
        }
    }
}

...
services.AddMvc(options => options.Filters.Add(new MyCleanupFilter());
```

This way your model gets to decide how to do the cleanup, and your filter just acts as a hook.

More to the point, is there something we're not doing in model binding that you think is 'cleanup' we should be doing? Is there a missing feature or setting here?

**Comment by mrahhal** [comment]
Oh, I see your point here. This is a nice solution to the problem without there being support from model binding itself.

The thing that made me think about this is when I learned about `IValidatableObject` as a way to run custom validation on the model itself. So naturally I thought that there's a similar interface that would let me run code on the model after model binding, which of course I didn't find. So I'd call it a missing feature (maybe a not so important one), because it naturally came to mind.

But your solution would suffice for the use case I presented 👍
